### PR TITLE
Fix bugs in the state trie

### PR DESCRIPTION
**Pull request by ultratwo**
### What was wrong?

The state trie incorrectly handling cases involving empty accounts and storage resulting in incorrect state roots. Most state roots involving storage will be wrong prior to this fix.

Code (mostly involving gas costs) that needs to know about the difference between `EMPTY_ACCOUNT` and a non-existent one will need to use the new function `get_account_optional()`.

#### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: src/ethereum/frontier/trie.py [hunk 1]
```diff
@@ -141,10 +147,12 @@ def encode_node(node: Node, storage_root: Optional[Bytes] = None) -> Bytes:
     if isinstance(node, Account):
         assert storage_root is not None
         return rlp.encode_account(node, storage_root)
-    elif isinstance(node, (Transaction, Receipt)):
+    elif isinstance(node, (Transaction, Receipt, U256)):
         return rlp.encode(cast(rlp.RLP, node))
+    elif isinstance(node, Bytes):
+        return node
     else:
-        return cast(Bytes, node)
```
Comment (voith) [inline src/ethereum/frontier/trie.py:147]: maybe raise an exception here: 
```python
raise NotImplementedError(f"encoding for {type(node)} is not currently implemented")
```

File: src/ethereum/frontier/trie.py [hunk 2]
```diff
@@ -141,10 +147,12 @@ def encode_node(node: Node, storage_root: Optional[Bytes] = None) -> Bytes:
...
+        assert False  # Unrecognized node type
```
Comment (voith) [inline src/ethereum/frontier/trie.py:155]: maybe raise an exception here: 
```python
raise NotImplementedError(f"encoding for {type(node)} is not currently implemented")
```

Comment (ultratwo) [reply src/ethereum/frontier/trie.py]: Fair

**Review by voith — APPROVED** [review]
This fixes all the bugs that I encountered while implementing pre-compiles.

**Review by SamWilsn — APPROVED** [review]

**Pull request merged**